File: backend-v2/src/zhishi/server/routes/goals.py
# src/zhishi/server/routes/goals.py
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel
from sqlalchemy.orm import Session
from zhishi.domain.goals import service
from zhishi.domain.goals.schemas import GoalCreate, KeyResultCreate
from zhishi.server.deps import get_db

router = APIRouter(prefix="/api/goals", tags=["goals"])


# ---- typed 响应（re #B1：openapi 从空 schema 变 $ref，字段与实际返回形状一致） ----

class KeyResultOut(BaseModel):
    id: int
    goal_id: int
    title: str
    kind: str                       # manual/tag_task_count/habit_checkins
    target_value: float
    current_value: float
    unit: str
    link: str


class GoalOut(BaseModel):
    id: int
    title: str
    notes: str
    status: str                     # active/paused/done/archived
    start_date: str | None = None   # YYYY-MM-DD
    end_date: str | None = None
    key_results: list[KeyResultOut]
    deleted_at: datetime | None = None   # 回收站语义（re #B2）：已删项透出删除时间


class GoalProgressItemOut(BaseModel):
    """progress 端点条目：自动类 KR 实时计算的 current_value + 0-100 整数进度。"""
    kr_id: int
    title: str
    kind: str
    target_value: float
    current_value: float
    unit: str
    progress: int


@router.post("", status_code=201, response_model=GoalOut)
def create_goal(payload: GoalCreate, db: Session = Depends(get_db)):
    return _goal_dict(service.create_goal(db, payload))


@router.get("", response_model=list[GoalOut])
def list_goals(include_deleted: bool = Query(False, description="包含已删除（回收站）目标"),
               include_archived: bool | None = Query(
                   None, deprecated=True,
                   description="旧参数名（re #B2 改名），等价 include_deleted，兼容保留"),
               db: Session = Depends(get_db)):
    include = include_deleted or bool(include_archived)
    return [_goal_dict(g) for g in service.list_goals(db, include_deleted=include)]


@router.get("/trash", response_model=list[GoalOut])
def list_trash(db: Session = Depends(get_db)):
    """回收站：软删目标列表（含 key_results 与 deleted_at，re #B2）。"""
    return [_goal_dict(g) for g in service.list_trash(db)]


@router.get("/{goal_id}", response_model=GoalOut)
def get_goal(goal_id: int, db: Session = Depends(get_db)):
    try:
        return _goal_dict(service.get_goal(db, goal_id))
    except LookupError:
        raise HTTPException(404, "目标不存在")


@router.patch("/{goal_id}", response_model=GoalOut)
def update_goal(goal_id: int, body: dict, db: Session = Depends(get_db)):
    try:
        return _goal_dict(service.update_goal(db, goal_id, **body))
    except LookupError:
        raise HTTPException(404, "目标不存在")


@router.delete("/{goal_id}", status_code=204)
def delete_goal(goal_id: int, db: Session = Depends(get_db)) -> None:
    try:
        service.delete_goal(db, goal_id)
    except LookupError:
        raise HTTPException(404, "目标不存在")


@router.post("/{goal_id}/restore", response_model=GoalOut)
def restore_goal(goal_id: int, db: Session = Depends(get_db)):
    try:
        return _goal_dict(service.restore_goal(db, goal_id))
    except LookupError:
        raise HTTPException(404, "目标不存在")


@router.delete("/{goal_id}/purge", status_code=204)
def purge_goal(goal_id: int, db: Session = Depends(get_db)) -> None:
    """硬删目标（级联 KR）。仅回收站中的可 purge：未软删 → 409。"""
    try:
        service.purge_goal(db, goal_id)
    except LookupError:
        raise HTTPException(404, "目标不存在")
    except ValueError as ex:
        raise HTTPException(409, str(ex))


@router.post("/{goal_id}/key-results", status_code=201, response_model=KeyResultOut)
def add_key_result(goal_id: int, payload: KeyResultCreate, db: Session = Depends(get_db)):
    try:
        kr = service.add_key_result(db, goal_id, payload)
    except LookupError:
        raise HTTPException(404, "目标不存在")
    return _kr_dict(kr)


@router.patch("/key-results/{kr_id}", response_model=KeyResultOut)
def update_key_result(kr_id: int, body: dict, db: Session = Depends(get_db)):
    try:
        if "current_value" in body and body["current_value"] is not None:
            kr = service.update_kr_progress(db, kr_id, current_value=body["current_value"])
        else:
            kr = db_get_kr(db, kr_id)
            if "title" in body and body["title"]:
                kr.title = body["title"]
            db.commit()
            db.refresh(kr)
    except LookupError:
        raise HTTPException(404, "关键结果不存在")
    return _kr_dict(kr)


@router.delete("/key-results/{kr_id}", status_code=204)
def delete_key_result(kr_id: int, db: Session = Depends(get_db)) -> None:
    try:
        service.delete_key_result(db, kr_id)
    except LookupError:
        raise HTTPException(404, "关键结果不存在")


@router.get("/{goal_id}/progress", response_model=list[GoalProgressItemOut])
def goal_progress(goal_id: int, db: Session = Depends(get_db)):
    try:
        return service.goal_progress(db, goal_id)
    except LookupError:
        raise HTTPException(404, "目标不存在")


def db_get_kr(db: Session, kr_id: int):
    from zhishi.domain.models import KeyResult
    kr = db.get(KeyResult, kr_id)
    if kr is None:
        raise LookupError(f"kr {kr_id} 不存在")
    return kr


def _goal_dict(g):
    return {"id": g.id, "title": g.title, "notes": g.notes, "status": g.status,
            "start_date": g.start_date.isoformat() if g.start_date else None,
            "end_date": g.end_date.isoformat() if g.end_date else None,
            "key_results": [_kr_dict(k) for k in g.key_results],
            "deleted_at": g.deleted_at.isoformat() if g.deleted_at else None}


def _kr_dict(k):
    return {"id": k.id, "goal_id": k.goal_id, "title": k.title, "kind": k.kind,
            "target_value": k.target_value, "current_value": k.current_value,
            "unit": k.unit, "link": k.link}

File: frontend-v2/app/src/api/goals.ts
/** 目标与关键结果 REST 接口。支持软删除、恢复和永久删除。 */
import type { components } from './contracts/rest'
import { http } from './http'

type schemas = components['schemas']

/** 目标状态（update_goal 文档：active/paused/done/archived）。 */
export type GoalStatus = 'active' | 'paused' | 'done' | 'archived'

/** 关键结果 = 生成 KeyResultOut（逐字段一致）。 */
export type KeyResult = schemas['KeyResultOut']

/** 目标 = 生成 GoalOut（含回收站字段 deleted_at；start/end_date 生成面可选）。 */
export type Goal = schemas['GoalOut']

/** 创建入参 = 生成 GoalCreate；notes 后端有缺省 → 保持可选。 */
export type GoalCreateInput = Partial<Omit<schemas['GoalCreate'], 'title'>> & { title: string }

/**
 * 更新入参：PATCH /api/goals/{id} 请求体在 openapi 仍为内联 object（未 typed）——
 * 手写面维持（update_goal 文档：title/notes/status/start_date/end_date 部分更新）。
 */
export interface GoalUpdateInput {
  title?: string
  notes?: string
  status?: GoalStatus
  start_date?: string | null
  end_date?: string | null
}

/** KR 创建入参 = 生成 KeyResultCreate；默认值字段保持可选。 */
export type KeyResultCreateInput = Partial<Omit<schemas['KeyResultCreate'], 'title'>> & { title: string }

/** progress 端点条目 = 生成 GoalProgressItemOut（0–100 整数进度）。 */
export type GoalProgressItem = schemas['GoalProgressItemOut']

/** 列表（改名 include_deleted；后端对旧名 include_archived 兼容保留）。 */
export function listGoals(includeDeleted = false): Promise<Goal[]> {
  return http.get<Goal[]>('/api/goals', { include_deleted: includeDeleted || undefined })
}

export function getGoal(goalId: number): Promise<Goal> {
  return http.get<Goal>(`/api/goals/${goalId}`)
}

export function createGoal(input: GoalCreateInput): Promise<Goal> {
  return http.post<Goal>('/api/goals', input)
}

export function updateGoal(goalId: number, patch: GoalUpdateInput): Promise<Goal> {
  return http.patch<Goal>(`/api/goals/${goalId}`, patch)
}

export function deleteGoal(goalId: number): Promise<void> {
  return http.del(`/api/goals/${goalId}`)
}

export function createKeyResult(goalId: number, input: KeyResultCreateInput): Promise<KeyResult> {
  return http.post<KeyResult>(`/api/goals/${goalId}/key-results`, input)
}

export function updateKeyResult(
  krId: number,
  patch: { title?: string; current_value?: number; target_value?: number; unit?: string },
): Promise<KeyResult> {
  return http.patch<KeyResult>(`/api/goals/key-results/${krId}`, patch)
}

export function deleteKeyResult(krId: number): Promise<void> {
  return http.del(`/api/goals/key-results/${krId}`)
}

export function getGoalProgress(goalId: number): Promise<GoalProgressItem[]> {
  return http.get<GoalProgressItem[]>(`/api/goals/${goalId}/progress`)
}
